include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -0,0 +1,337 @@
+namespace asset
+{
+
+class ICPUVirtualTexture final : public IVirtualTexture<ICPUImageView>
+{
+    using base_t = IVirtualTexture<ICPUImageView>;
+
+public:
+    class ICPUVTResidentStorage final : public base_t::IVTResidentStorage
+    {
+        using base_t = base_t::IVTResidentStorage;
+
+    public:
+        ICPUVTResidentStorage(E_FORMAT _format, uint32_t _extent, uint32_t _layers, uint32_t _tilesPerDim) :
+            base_t(_layers, _tilesPerDim)
+        {
+            ICPUImage::SCreationParams params;
+            params.extent = {_extent,_extent,1u};
+            params.format = _format;
+            params.arrayLayers = _layers;
+            params.mipLevels = 1u;
+            params.type = IImage::ET_2D;
+            params.samples = IImage::ESCF_1_BIT;
+            params.flags = static_cast<IImage::E_CREATE_FLAGS>(0);
+
+            image = ICPUImage::create(std::move(params));
+            {
+                auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<IImage::SBufferCopy>>(1ull);
+                auto& region = regions->front();
+                region.imageSubresource.mipLevel = 0u;
+                region.imageSubresource.baseArrayLayer = 0u;
+                region.imageSubresource.layerCount = _layers;
+                region.bufferOffset = 0u;
+                region.bufferRowLength = _extent;
+                region.bufferImageHeight = 0u; //tightly packed
+                region.imageOffset = {0u,0u,0u};
+                region.imageExtent = params.extent;
+                auto buffer = core::make_smart_refctd_ptr<ICPUBuffer>(getTexelOrBlockBytesize(_format) * params.extent.width*params.extent.height*params.arrayLayers);
+                image->setBufferAndRegions(std::move(buffer), regions);
+            }
+        }
+
+    private:
+        core::smart_refctd_ptr<ICPUImageView> createView_internal(ICPUImageView::SCreationParams&& _params) const override
+        {
+            return ICPUImageView::create(std::move(_params));
+        }
+    };
+
+    ICPUVirtualTexture(
+        const base_t::IVTResidentStorage::SCreationParams* _residentStorageParams,
+        uint32_t _residentStorageCount,
+        uint32_t _pgTabSzxy_log2 = 8u,
+        uint32_t _pgTabLayers = 256u,
+        uint32_t _pgSzxy_log2 = 7u,
+        uint32_t _tilePadding = 9u,
+        uint32_t _maxAllocatableTexSz_log2 = 14u
+    ) : IVirtualTexture(
+        _pgTabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _tilePadding
+    ) {
+        m_pageTable = createPageTable(_pgTabSzxy_log2, _pgTabLayers, _pgSzxy_log2, _maxAllocatableTexSz_log2);
+        initResidentStorage(_residentStorageParams, _residentStorageCount);
+    }
+
+    STextureData pack(const ICPUImage* _img, const IImage::SSubresourceRange& _subres, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv, ISampler::E_TEXTURE_BORDER_COLOR _borderColor) override
+    {
+        const E_FORMAT format = _img->getCreationParameters().format;
```

Comment:
you should only be getting the format class not using the format directly, the image we're packing could have views that we intend to use which are completely different than the format

Reply:
See https://github.com/buildaworldnet/IrrlichtBAW/pull/474/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=No+extension&hide-deleted-files=true#r418064940